app.ts

```diff
@@ -0,0 +1,153 @@
+
+	// create table
+	let table = document.createElement("table");
+	table.className = "gridtable";
+    let thead = document.createElement("thead");
+    let tbody = document.createElement("tbody");
+	let headRow = document.createElement("tr");
+
+	// create column headers
+    for(const header of columns) {
+		var th = document.createElement("th");
+		th.appendChild(document.createTextNode(header));
+		headRow.appendChild(th);
+    };
+    thead.appendChild(headRow);
+	table.appendChild(thead);
+
+	// create rows
+    records.forEach(function (el) {
+		let tr = document.createElement("tr");
+		for (var o in el) {
+			let td = document.createElement("td");
+			td.appendChild(document.createTextNode(el[o]))
+			tr.appendChild(td);
+		}
+		tbody.appendChild(tr);
+	});
+	table.appendChild(tbody);
+
+	// builds footer elements (buttons, search field)
+	document.getElementById("footer").innerHTML = `<a onclick="goPrevious()">&laquo;</a>
+	<a onclick="goNext()">&raquo;</a>
+	<input type="text" id="search" onkeyup="search(event)" placeholder="Enter ID...">`;
+
+	document.getElementById("loader").remove();
+
+    return table;
+}
+
+// filters rows by user input ID
+function search(e){
+	let input = $("#search").val().toString();
+	let $table = $("tbody tr").toArray();
+
+	for(const record of $table){
+		let td = record.getElementsByTagName("td")[0];
+		record.style.display="block";
+		if(td){
+			let val = td.innerText || td.textContent;
+			if(val.indexOf(input) > -1){
+				record.style.display = "";
+			}
+			else{
+				record.style.display = "none";
+			} 		
+		}
+	}
+
+	// if record doesn't exist on current page, query database and display results
+	if(e.which == 13)
+	{
+		let end = parseInt(input) + calculateRows();
+		clearTable();
+		loadPage(parseInt(input), end);
+	}
+}
+
+// load the previous page
+function goPrevious(){
+	let numRows = calculateRows();
+	if(lastIndex > numRows){
+		clearTable();
+		loadPage(firstIndex-numRows, firstIndex-1);
+	}
+}
+
+// load the next page
+function goNext(){
+	if(lastIndex < recordCount-1)
+	{
+		clearTable();
+		loadPage(lastIndex+1, lastIndex+calculateRows());
+	}
+}
+
+// clears the table and footer elements
+function clearTable(){
+	document.getElementById("content").innerHTML="";
+	document.getElementById("footer").innerHTML="";
+}
+
+// loads the columns and rows to be displayed based on start and end row indices
+function loadPage(start:number, end:number){
+	firstIndex = start;
+	lastIndex = end;
+
+	// check bounds
+	if(firstIndex<0) firstIndex = 0;
+	if(lastIndex>=recordCount) lastIndex = recordCount-1;
+	if(firstIndex>=recordCount){
+		firstIndex = recordCount - calculateRows();
+	} 
+
+	// create loader
+	let loader = document.createElement("loader");
+	document.getElementById("content").appendChild(loader);
+	loader.innerHTML=`<div class="loader" id="loader"></div>`;
+	
+	// outter function to fetch column headers from server
+	$.get("http://localhost:2050/columns", function(columns){
```

Comment:
This call will always return the same values, so you can call it once and cache the results, like you did with record count.